test/envtest/admission_webhook_envtest_test.go

```diff
@@ -0,0 +1,213 @@
+					},
+				},
+			},
+			expectErrorContains: `vault configuration in invalid: schema violation (name: vault 'env1' is not installed)`,
+		},
+		{
+			name: "should fail the validation if the spec.config does not pass the schema check of Kong gateway",
+			kongVault: &kongv1alpha1.KongVault{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "vault-invalid-config",
+					Annotations: map[string]string{
+						annotations.IngressClassKey: annotations.DefaultIngressClass,
+					},
+				},
+				Spec: kongv1alpha1.KongVaultSpec{
+					Backend:     "env",
+					Prefix:      "invalid-config",
+					Description: "test env vault",
+					Config: apiextensionsv1.JSON{
+						Raw: []byte(`{"prefix":"kong-env-test","foo":"bar"}`),
+					},
+				},
+			},
+			expectErrorContains: `vault configuration in invalid: schema violation (config.foo: unknown field)`,
+		},
+		{
+			name: "should fail the validation if spec.prefix is duplicate",
+			kongVault: &kongv1alpha1.KongVault{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "vault-dupe",
+					Annotations: map[string]string{
+						annotations.IngressClassKey: annotations.DefaultIngressClass,
+					},
+				},
+				Spec: kongv1alpha1.KongVaultSpec{
+					Backend:     "env",
+					Prefix:      prefixForDuplicationTest, // This is the same prefix as the one created in setup.
+					Description: "test env vault",
+				},
+			},
+			expectErrorContains: fmt.Sprintf(`spec.prefix "%s" is duplicate with existing KongVault`, prefixForDuplicationTest),
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			err := ctrlClient.Create(ctx, tc.kongVault)
+			if tc.expectErrorContains != "" {
+				require.Error(t, err)
+				require.Contains(t, err.Error(), tc.expectErrorContains)
+			} else {
+				require.NoError(t, err)
+			}
+		})
+	}
+}
+
+// prepareKongVaultAlreadyProgrammedInGateway creates a KongVault and waits until it gets programmed in Gateway.
+func prepareKongVaultAlreadyProgrammedInGateway(
+	ctx context.Context,
+	t *testing.T,
+	ctrlClient client.Client,
+	vaultPrefix string,
+) {
+	name := uuid.NewString()
+	require.NoError(t, ctrlClient.Create(ctx, &kongv1alpha1.KongVault{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+			Annotations: map[string]string{
+				annotations.IngressClassKey: annotations.DefaultIngressClass,
+			},
+		},
+		Spec: kongv1alpha1.KongVaultSpec{
+			Backend:     "env",
+			Prefix:      vaultPrefix,
+			Description: "vault description",
+		},
+	}))
+
+	require.Eventually(t, func() bool {
+		kv := &kongv1alpha1.KongVault{}
+		err := ctrlClient.Get(ctx, k8stypes.NamespacedName{Name: name}, kv)
+		if err != nil {
+			t.Logf("Failed to get KongVault: %v", err)
+			return false
+		}
+		programmed, ok := lo.Find(kv.Status.Conditions, func(c metav1.Condition) bool {
+			return c.Type == "Programmed"
+		})
+		if !ok || programmed.Status != metav1.ConditionTrue {
+			t.Log("KongVault created for duplicate prefix check has not been programmed yet")
```

Comment:
- Might as well include the name of the problematic kongvault.
- Wouldn't this be too verbose given the 20ms interval? 🤔

Reply:
👍 Yeah, let's just log the starting point of waiting: https://github.com/Kong/kubernetes-ingress-controller/pull/5605/commits/28d88477c7272d422c6c2b79c011a460f7536d94